Add dummy RPCs to Authenticate and ChangePassword

0xB10C opened the pull request:
This PR adds two dummy RPCs to authenticate and change the password. This shall only be used for testing and development. 

See https://github.com/digitalbitbox/bitbox-base/pull/164#issuecomment-526148566 for current behavior.

Tomasvrba commented on middleware/src/middleware.go:
```diff
@@ -251,6 +251,26 @@ func (middleware *Middleware) Restore(method rpcmessages.RestoreArgs) (rpcmessag
+		"djb":     "shift9",
+	}
+
+	if expectedPasssword, userIsInMap := dummyUsers[args.Username]; userIsInMap {
```
Personal preference to use a custom variable instead of just `ok` which is generally the case in Go?
I don't really mind either way, just wondering :)

0xB10C replied:
Had `ok` at first, but wanted the code to be clear(er), when dealing with passwords. But your right, I'd normally expect a `ok` as well.

Stadicus commented:
I also think that we should already have the insecure factory passwort built in. I think we also discussed to use `userAuthenticate` and `userChangePassword`.

Our development tests with the app could be:
1. App tries `userAuthenticate` with admin/1234
1. Success, Base is not yet initialized! --> set new password
1. Tester sets to satoshi/shift1 (or any of the above) with `userChangePassword`
1. App is closed and restarted. User authenticates with mallory/evilhack --> Auth fails!
1. App is closed and restarted. User authenticates with satoshi/shift1, sets "remember password"
1. App is closed and restarted. App automatically authenticates with satoshi/shift1

As it is an insecure password, I'd rather have it obiously insecure than to have something cryptic that only seems secure. What do you think about 1234, bitboxbase or CanIHazPassword?

Only change necessary are RPC naming and inclusion of the standard password.

Tomasvrba commented:
> As it is an insecure password, I'd rather have it obiously insecure than to have something cryptic that only seems secure. What do you think about 1234, bitboxbase or CanIHazPassword?
> 

I definitely vote for "ICanHasPassword?"

0xB10C commented:
Good feedback, should have looked closer at the actual spec.

0xB10C commented:
Addressed feedback and added a dummy state as discussed with @Stadicus. Current behavior is as follows:

If the Base is not set up (tracked as the `dummyIsBaseSetup` boolean and reset on middleware restart) you can *only* `UserAuthenticate` with:

| username | password |
|----------|----------|
| admin  | ICanHasPassword?  |

Calling the RPC `UserChangePassword` with the username `admin` let's you choose a new admin password (this is saved in the `dummyAdminPassword` string and reset on middleware restart as well). Changing the admin password sets `dummyIsBaseSetup = true`. It's now possible to login with one of these three:

| username | password |
|----------|----------|
| admin  | (user-set-password)   |
| satoshi     | shift1   |
| dev      | dev  |

The user set password must be at least 8 chars. This can be used to test error behavior when implementing the `userChangePassword` functionality in the front end.

0xB10C commented:
Rebased onto of #169 and refactored the RPC to return `ErrorResponse` as discussed. Currently not returning an (error) code, since it's a dummy RPC implementation.

0xB10C commented:
Rebased onto master now.

Tomasvrba commented on middleware/src/middleware.go:
```diff
@@ -292,10 +310,10 @@ func (middleware *Middleware) UserChangePassword(args rpcmessages.UserChangePass
 
-	errorMessage := "authentication unsuccessful"
-	return rpcmessages.GenericResponse{Success: false, Message: errorMessage}, errors.New(errorMessage)
+	return rpcmessages.ErrorResponse{Success: false, Message: "password change unsuccessful (to short)"}
```
Typo, "too short"

Tomasvrba commented on middleware/src/middleware.go:
```diff
@@ -292,10 +310,10 @@ func (middleware *Middleware) UserChangePassword(args rpcmessages.UserChangePass
 
-	errorMessage := "authentication unsuccessful"
-	return rpcmessages.GenericResponse{Success: false, Message: errorMessage}, errors.New(errorMessage)
+	return rpcmessages.ErrorResponse{Success: false, Message: "password change unsuccessful (to short)"}
```
You can also get an ErrorResponse when the provided username is not "admin" right? In that case the message would be incorrect.

0xB10C replied:
No, the `if len(args.NewPassword) >= 8 {...}` block is closed at this point :)

Tomasvrba replied:
Yep, makes sense. Maybe I looked at a previous commit before or just read it wrong :+1:

0xB10C commented:
Squashed and fixed typo.
